ext/standard/var.c

```diff
@@ -1406,6 +1406,11 @@ PHPAPI void php_unserialize_with_options(zval *return_value, const char *buf, co
 			zval_ptr_dtor(return_value);
 		}
 		RETVAL_FALSE;
+	} else if ((char*)p < buf + buf_len) {
```

Comment:
Convert this to an `else {` then a standalone if and always execute the level+refcounted checks that were executed before? I think we still want to copy retval to return_value right now when unserialize triggers another call to unserialize (e.g. class implementation of `::__serialize`) calling `\unserialize()` with the serialization of an array with extra data), and dump the unserialized result

Add a test that this works properly with that - I expect it's broken due to the lack of `ZVAL_COPY(return_value, retval)`

```c
	if (BG(unserialize).level > 1) {
		retval = var_tmp_var(&var_hash);
	} else {
		retval = return_value;
	}
```

```c
if (BG(unserialize).level > 1) {
		ZVAL_COPY(return_value, retval);
	} else if (Z_REFCOUNTED_P(return_value)) {
		zend_refcounted *ref = Z_COUNTED_P(return_value);
		gc_check_possible_root(ref);
	}
} /* End the new else */
```

Are you planning to remove this in a subsequent PHP release? E_DEPRECATED is another option, though it might be better to wait until we know what types of setups are effected (e.g. is anything storing data with trailing null bytes, are any extensions calling it with a length off by one, etc)

Reply:
> Convert this to an else { then a standalone if and always execute the level+refcounted checks that were executed before?

Indeed. The diff looked fishy to me, but I could not pinpoint what was wrong with it. You were right in that it breaks nested `unserialize()` calls (though only for `Serializable`).

Reply:
> Are you planning to remove this in a subsequent PHP release?

I'm not currently planning anything specifically. I came across this, while working with the code and I wasted to keep track of it, so I created a PR.

> To me, E_WARNING makes some sense for existing unserialize() failures because it warns that unserialize is not returning data (though it'll start throwing later), but for this, E_DEPRECATED might make sense for something that works in 8.x and older but will stop working in 9

`unserialize()` will also emit E_WARNING when it successfully returns data, but the data may be malformed:

```
<?php

echo file_get_contents(__FILE__), PHP_EOL, "====================", PHP_EOL;

var_dump(unserialize('i:111111111111111111111111111111111;'));

====================

Warning: unserialize(): Numerical result out of range in /home/timwolla/Schreibtisch/php-src/test.php on line 5
int(9223372036854775807)
```

I believe this is similar: unserialize is able to make sense of the input, but not all of the input. Thus the operation might be “lossy”. E_WARNING seems fitting.

Reply:
> Indeed. The diff looked fishy to me, but I could not pinpoint what was wrong with it. You were right in that it breaks nested unserialize() calls (though only for Serializable).

That looks right - I wasn't familiar with the purpose of `BG(serialize_lock)`. I took another look at the implementation and it's deliberately incremented before calling `__wakeup`, `__sleep`, `__serialize`, `__unserialize`, autoloading, and unserialize_callback_func from the optional ini setting (e.g. everything except for the legacy Serializable, I'd believe)

Looking into it just to see if any additional tests would be needed for this functionality in php 9.0 - the existing tests should be fine

So it'd probably be safe to remove `serialize_lock` in php 9.0 if the Serializable functionality is removed for internal classes, which the rfc mentioned we planned to do
https://wiki.php.net/rfc/phase_out_serializable

> In PHP 9.0 the Serializable interface will be removed and **unserialize() will reject payloads using the C serialization format.** Code needing to support both PHP < 7.4 and PHP >= 9.0 may polyfill the Serializable interface, though it will have no effect on serialization.

Reply:
> I believe this is similar: unserialize is able to make sense of the input, but not all of the input. Thus the operation might be “lossy”. E_WARNING seems fitting.

Oh, I missed that this is still returning the data (I was distracted by the bug previously failing to set the return_value). In that case, I'd agree with making it an E_WARNING